MDEV-17467 Add regression function

Weijun-H opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling this template <3

If you have any questions related to MariaDB or you just want to
hang out and meet other community members, please join us on
https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue
that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-17467*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed, what was it looking like before
   the change and how it's looking with this patch applied
3. Do you think this patch might introduce side-effects in
   other parts of the server?
-->
## Description
The following linear regression functions will be created:
- REGR_SLOPE
- REGR_INTERCEPT
- REGR_COUNT
- REGR_R2
- REGR_AVGX
- REGR_AVGY
- REGR_SXX
- REGR_SYY
- REGR_SXY
## How can this PR be tested?
I have created the test file  in `main/regression_function`, but it sill is not comprehensive.
<!--
TODO: modify the automated test suite to verify that the PR causes MariaDB to
behave as intended. Consult the documentation on
["Writing good test cases"](https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server).
In many cases, this will be as simple as modifying one `.test` and one `.result`
file in the `mysql-test/` subdirectory. Without _automated_ tests, future regressions
in the expected behavior can't be automatically detected and verified.

If the changes are not amenable to automated testing, please explain why not and
carefully describe how to test manually.
-->

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

<!--
You might consider answering some questions like:
1. Does this affect the on-disk format used by MariaDB?
2. Does this change any behavior experienced by a user
   who upgrades from a version prior to this patch?
3. Would a user be able to start MariaDB on a datadir
   created prior to your fix?
-->

Weijun-H committed:
MDEV-17467: add regression functions

Weijun-H commented:
@cvicentiu

cvicentiu commented on mysql-test/main/regression_function.test:
```diff
@@ -0,0 +1,126 @@
+#
+
+--disable_warnings
+DROP TABLE IF EXISTS t1, t2;
```
This line isn't necessary. Assume test cases always start with a clean environment.

cvicentiu commented on sql/item.h:
```diff
@@ -873,10 +873,11 @@ class Item :public Value_source,
-             SUBSELECT_ITEM, ROW_ITEM, CACHE_ITEM, TYPE_HOLDER,
-             PARAM_ITEM, TRIGGER_FIELD_ITEM,
-             EXPR_CACHE_ITEM};
+             FIELD_REGR_AVGX_ITEM, FIELD_REGR_AVGY_ITEM, FIELD_REGR_SXX_ITEM,
```
Let's try this a bit differently. You don't need a different field type for each individual Item.

Declare only one, `FIELD_REGR_ITEM` and write code according to that. See how it turns out. I don't think you actually need a class for each type here as you're effectively always storing and retrieving Regr_transition data structures.

cvicentiu reviewed (changes requested):
Alright WeiJun. First of all, really good job coding this. I have given you some comments in various places in your patch. You don't *have* to agree to all of them. So if there's anything you find strange, let's discuss. :)

cvicentiu commented:
Hi @Weijun-H!

I didn't get a chance to do an in-depth review yet. I'll do it as soon as I can.

However here's the thing I would like you to extend moving further:

It would be ideal if we could implement "removal" of a row from a regression set. This happens with `regr_sxx(a, b) over (rows between 2 preceding and 2 following)`
Here we effectively calculate a regression of 5 elements. For performance reasons we would like our window functions to support removal. Relevant areas of the code are in:
sql/sql_window.cc, and the functions, `Item_sum::supports_removal()` and `Item_sum::remove()`. You can take `Item_sum_count::remove()` and `Item_sum_count::supports_removal()` as an example.

See if you can implement those for our regression functions.

Another thing to tackle:
Test cases with views, prepared statements, replication to a slave (queries like `create table t1 select regr_sxx(...)`) for instance.

Weijun-H commented:
> Hi @Weijun-H!
> 
> I didn't get a chance to do an in-depth review yet. I'll do it as soon as I can.
> 
> However here's the thing I would like you to extend moving further:
> 
> It would be ideal if we could implement "removal" of a row from a regression set. This happens with `regr_sxx(a, b) over (rows between 2 preceding and 2 following)` Here we effectively calculate a regression of 5 elements. For performance reasons we would like our window functions to support removal. Relevant areas of the code are in: sql/sql_window.cc, and the functions, `Item_sum::supports_removal()` and `Item_sum::remove()`. You can take `Item_sum_count::remove()` and `Item_sum_count::supports_removal()` as an example.
> 
> See if you can implement those for our regression functions.
> 
> Another thing to tackle: Test cases with views, prepared statements, replication to a slave (queries like `create table t1 select regr_sxx(...)`) for instance.

Hi @cvicentiu. This extention sounds very interesting, please give me some time to research and implement them.

Weijun-H committed:
complete regresion function in OVER BY preceding and following

Weijun-H committed:
add tests for view

Weijun-H committed:
add prepared statements test

Weijun-H committed:
CURRENT_TEST: main.regression_function_prepare_error
mysqltest: At line 60: query 'SET @col1 =age, @col2 = weight' failed:
ER_BAD_FIELD_ERROR (1054): Unknown column 'age' in 'field list'

cvicentiu commented on sql/item_sum.cc:
```diff
@@ -2197,6 +2216,340 @@ double Stddev::result(bool is_sample_variance)
+    sxy= sxy - diff * (nr_y - syy.get_m());
+    syy.remove(nr_y);
+    N--;
+  }
```
Same as a previous comment. Let's use a DBUG_ASSERT to ensure we never get called with N == 0.

cvicentiu commented on sql/item_sum.cc:
```diff
@@ -2197,6 +2216,340 @@ double Stddev::result(bool is_sample_variance)
+{
+  DBUG_ENTER("Item_sum_regr_count::val_int");
+  DBUG_ASSERT(fixed());
+  // if (aggr)
```
Lets remove commented out code.

cvicentiu commented on sql/item_sum.cc:
```diff
@@ -2387,6 +2755,91 @@ void Item_sum_min_max::clear()
 
+double Item_regr_sxx_field::val_real()
+{
+  // fix_fields() never calls for this Item
```
Hmm, that's odd. I'm going to debug this and see why that'st he case.

Hinal-Srivastava commented:
Hello,
I am interested in contributing to this project. Can you please tell me some resources to get started and also guide me towards some first timer tasks? 
Thank you! Looking forward to hearing from you

gkodinov commented:
Hello,

Is there an active developer willing to work on this PR?
If there we can "drive this home" by:

1. Work on https://jira.mariadb.org/browse/MDEV-17467 so we turn it into a complete specification document
2. Solicit a go-ahead on the feature from the MariaDB captains.
3. Rebase the PR to the latest trunk
4. Submit the updated PR for a preliminary review to me
5. We solicit final review from the MariaDB captains

Anybody willing to work with me on this?

mohitbalwani commented:
Hello @gkodinov 👋
I am really interested to complete the remaining work of this PR. I have started looking into the existing work and exploring it. I am planning to follow up with you on it in Zulip channel.

gkodinov commented:
There has been no activity on this one since 13 Mar. Converting to draft. Please re-open when/if you resume working on it.
